src/utils.ts

```diff
@@ -1158,3 +1158,23 @@ export function hasAtomicOperators(doc: Document | Document[]): boolean {
   const keys = Object.keys(doc);
   return keys.length > 0 && keys[0][0] === '$';
 }
+
+export function deepFreeze<O extends Record<string, any>>(object: O): Readonly<O> {
+  // Retrieve the property names defined on object
+  const propNames = Object.getOwnPropertyNames(object);
+
+  // Freeze properties before freezing self
+
+  for (const name of propNames) {
```

Comment:
Concern here about the combination of using deepFreeze within getters, this for loop will run on every access `this.thing`, and it's pretty hefty for loop. We may be making the object immutable but it's not the same object every-time this getter is run. Thoughts on some memoization or caching functionality?

Reply:
This could easily be achieved by setting `this.cacheBuildOptions` from within the getter, and return that if it exists.

Reply:
I think it's fine in terms of performance if we minimize use of the getter, I commented about that above. In terms of immutability, I don't think we can expect something that is gradually built on an object to be the same over the object's lifespan, so I think this is OK.